ydb/core/tx/schemeshard/schemeshard_utils.cpp

```diff
@@ -350,36 +352,48 @@ void SetImplTablePartitionConfig(
 }
 
 void FillIndexImplTableColumns(
-    const THashMap<ui32, struct NSchemeShard::TTableInfo::TColumn>& baseTableColumns,
+    const auto& baseTableColumns,
     const TTableColumns& implTableColumns,
     NKikimrSchemeOp::TTableDescription& implTableDesc)
 {
+    const auto was = implTableDesc.ColumnsSize();
 
     THashMap<TString, ui32> implKeyToImplColumn;
     for (ui32 keyId = 0; keyId < implTableColumns.Keys.size(); ++keyId) {
         implKeyToImplColumn[implTableColumns.Keys[keyId]] = keyId;
     }
 
-    for (auto& iter: baseTableColumns) {
-        const NSchemeShard::TTableInfo::TColumn& column = iter.second;
-        if (column.IsDropped()) {
-            continue;
+    for (ui32 i = Max<ui32>() / 2; auto& iter: baseTableColumns) {
+        NKikimrSchemeOp::TColumnDescription* item = nullptr;
+        if constexpr (std::is_same_v<const std::pair<const ui32, NSchemeShard::TTableInfo::TColumn>&, decltype(iter)>) {
+            const auto& column = iter.second;
+            if (!column.IsDropped() && implTableColumns.Columns.contains(column.Name)) {
+                item = implTableDesc.AddColumns();
+                item->SetName(column.Name);
+                item->SetType(NScheme::TypeName(column.PType, column.PTypeMod));
+                item->SetNotNull(column.NotNull);
+            }
+        } else if constexpr (std::is_same_v<const NKikimrSchemeOp::TColumnDescription&, decltype(iter)>) {
+            if (implTableColumns.Columns.contains(iter.GetName())) {
+                item = implTableDesc.AddColumns();
+                *item = iter;
+                item->ClearFamily();
+                item->ClearFamilyName();
+                item->ClearDefaultValue();
+            }
+        } else {
+            static_assert(dependent_false<decltype(iter)>::value);
         }
-
-        if (implTableColumns.Columns.contains(column.Name)) {
-            auto item = implTableDesc.AddColumns();
-            item->SetName(column.Name);
-            item->SetType(NScheme::TypeName(column.PType, column.PTypeMod));
-            item->SetNotNull(column.NotNull);
-            ui32 order = Max<ui32>();
-            if (implKeyToImplColumn.contains(column.Name)) {
-                order = implKeyToImplColumn.at(column.Name);
+        if (item) {
+            ui32 order = i++;
+            if (const auto* id = implKeyToImplColumn.FindPtr(item->GetName())) {
+                order = *id;
             }
             item->SetId(order);
         }
     }
 
-    std::sort(implTableDesc.MutableColumns()->begin(),
+    std::sort(implTableDesc.MutableColumns()->begin() + was,
```

Comment:
I think that this column sorting/ordering thing (orders > Max<ui32>() / 2 and partial sorting) deserves an explanation to the reader.
And a test (that it actually correct in all cases).

Reply:
Added test and comments to explain this